dial.go

```diff
@@ -138,28 +141,44 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
 			return nil, err
 		}
 	}
+	log.Debugf("dialer %s selected protocol: %s", d, selectedProto)
+	logdial["protocol"] = selectedProto
 
 	c = newSingleConn(ctx, d.LocalPeer, remote, maconn)
 	if d.PrivateKey == nil || !iconn.EncryptConnections {
-		log.Warning("dialer %s dialing INSECURELY %s at %s!", d, remote, raddr)
+		if selectedProto != NoEncryptionTag {
+			panic("expected plaintext negotiation")
+		}
+		log.Warningf("dialer %s dialing INSECURELY %s at %s!", d, remote, raddr)
 		return c, nil
 	}
 
-	c2, err := newSecureConn(ctx, d.PrivateKey, c)
-	if err != nil {
-		c.Close()
-		return nil, err
+	switch selectedProto {
+	case TLSTag:
+		c2, err := newTLSConn(ctx, d.PrivateKey, c, true)
+		if err != nil {
+			c.Close()
+			return nil, err
+		}
+		c = c2
+	case SecioTag:
+		c2, err := newSecureConn(ctx, d.PrivateKey, c)
+		if err != nil {
+			c.Close()
+			return nil, err
+		}
+		c = c2
```

Comment:
probably want to throw a default in here, just in case, and also explicitly acknowledge the `NoEncryptionTag` in its own case.

Reply:
Done. Also renamed the tag to `TLS12Tag` since you wouldn't want to change an exported const.